[tata32000] iP

tata32000 opened the pull request:
# Meep

A chatbot + task manager

---

## Why Meep?

* **BEAUTIFUL** graphical user interface (GUI)
* Easy to learn and use
* **QUICK** loading and unloading of the saved tasks

---

## How to Use?

1. Download it from [here](https://github.com/tata32000/ip/releases/download/A-Jar/meep.jar).
1. Go to the directory with the downloaded file
1. Run the following command:
    ```bash
    java -jar Meep.jar
    ```
1. Start using!

---

## Features

* Managing things to do
* Managing events
* Managing deadlines
* Finding saved Tasks

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

tata32000 committed:
Rename main class

tata32000 committed:
Add greeting

tata32000 committed:
Add echo and exits

tata32000 committed:
Add store class to manage text list

tata32000 committed:
Rename store to taskList

tata32000 committed:
Rename items to tasks in TaskList.java

tata32000 committed:
Refactor main class and create Parser class

tata32000 committed:
Add mark and unmark to TaskList

tata32000 committed:
Refactor status icon for Task

tata32000 committed:
Add private final to tasks in TaskList.java

tata32000 committed:
Add AI.md

tata32000 committed:
Add Deadline and refactor printing methods

tata32000 committed:
Add todo class

tata32000 committed:
Change protected to private final in Deadline.java

tata32000 committed:
Add Event class

tata32000 committed:
Format addTask method description string

tata32000 committed:
Add this keyword to tasks

tata32000 committed:
Change getLast() method to get()

tata32000 committed:
Rename Duke to Meep in runtest.sh

tata32000 committed:
Add test cases

tata32000 committed:
Fix bug where todo keyword followed by any character is a valid command

tata32000 committed:
Add Deadline and Event error handling

tata32000 committed:
Update test cases

tata32000 committed:
Add error handling for mark and unmark command

tata32000 committed:
Update test cases

tata32000 committed:
Refactor command checking

tata32000 committed:
Add delete command

tata32000 committed:
Update test cases for delete command

tata32000 committed:
Refactor input waiting

tata32000 committed:
Add enum and rename Command to CommandParser

tata32000 committed:
Add FileUtils class to handle file save

tata32000 committed:
Add data.txt to gitignore file

tata32000 committed:
Add save format methods to Task and TaskList

tata32000 committed:
Add getSaveFormat method to Todo and Deadline

tata32000 committed:
Add loadTasks and saveTasks to Main class

tata32000 committed:
Refactor processTaskString method in FileUtils class

tata32000 committed:
Add getSaveFormat method to Event class

tata32000 committed:
Add try catch block to processTaskString method in FileUtils class

tata32000 committed:
Merge branch 'branch-Level-7' of https://github.com/tata32000/ip

tata32000 committed:
Update AI.md

tata32000 committed:
Add date formatter methods in Parser class

tata32000 committed:
Add invalidDateFormat method in FormattedPrint class

tata32000 committed:
Add date format for Event class parser

tata32000 committed:
Merge branch 'branch-Level-8'

tata32000 committed:
Rename FormattedPrint to Ui class and add error display methods

tata32000 committed:
Add new constructor in TaskList class

tata32000 committed:
Refactor Storage class to include constructor and Exception handling

tata32000 committed:
Add MeepException class

tata32000 committed:
Refactor Parser class to use Ui class

tata32000 committed:
Modify runtest.sh to delete data.txt from previous run

tata32000 committed:
Refactor Main method to use more OOP

tata32000 committed:
Modify test cases

tata32000 committed:
Remove unused method in TaskList class

tata32000 committed:
Refactor the classes into suitable java packages

tata32000 committed:
Modify runtest.sh to include all java files in the src folder

tata32000 committed:
Merge branch 'branch-packages'

tata32000 committed:
Remove unused method in Task.java

tata32000 committed:
Merge branch 'add-gradle-support'

tata32000 committed:
Add newline at end of file to runtest.sh

tata32000 committed:
Modify main class name in build.gradle file

tata32000 committed:
Add unit tests for CommandParser class

tata32000 committed:
Update AI.md

tata32000 committed:
Add unit test for TaskList class

tata32000 committed:
Update AI.md to include javadoc generation

tata32000 committed:
Add JavaDoc comments to Meep class

tata32000 committed:
Add JavaDoc comments to Task class

tata32000 committed:
Add JavaDoc comments to TaskList class

tata32000 committed:
Add JavaDoc comments to Storage class

tata32000 committed:
Add JavaDoc comments to MeepException class

tata32000 committed:
Add JavaDoc comments to CommandParser class

tata32000 committed:
Add JavaDoc comments to Parser class

tata32000 committed:
Add JavaDoc comments to Deadline class

tata32000 committed:
Add JavaDoc comments to Event class

tata32000 committed:
Add JavaDoc comments to Todo class

tata32000 committed:
Add JavaDoc comments to Ui class

tata32000 committed:
Rearrange the import order in Storage class

tata32000 committed:
Add find related methods to Ui class

tata32000 committed:
Add findTasks method to TaskList class

tata32000 committed:
Add find command to Parser class

tata32000 committed:
Fix findTasks method to match ignore case

tata32000 committed:
Modify runtest.sh

tata32000 committed:
Update test cases

tata32000 committed:
Merge branch 'branch-A-JavaDoc'

tata32000 committed:
Merge branch 'branch-A-CodingStandard'

tata32000 committed:
Merge branch 'branch-Level-9'

tata32000 committed:
Add JavaDoc for find related methods

tata32000 committed:
Add checkstyle config files

tata32000 committed:
Modify build.gradle to include checkstyle

tata32000 committed:
Fix coding style in Ui class

tata32000 committed:
Fix coding style in Meep class

tata32000 committed:
Fix coding style in Task class

tata32000 committed:
Fix coding style in Storage class

tata32000 committed:
Fix coding style in MeepException.java

tata32000 committed:
Fix coding style in CommandParser.java

tata32000 committed:
Fix coding style in Parser.java

tata32000 committed:
Fix coding style in CommandParserTest.java

tata32000 committed:
Fix coding style in TaskListTest.java

tata32000 committed:
Merge branch 'branch-A-CheckStyle'

QuekBingHeng commented on src/main/java/meep/Meep.java:
```diff
@@ -0,0 +1,65 @@
+    public void run() {
+        ui.greeting();
+        Scanner scanner = new Scanner(System.in);
+        boolean isDone = false;
```
Nice naming convention for boolean!

QuekBingHeng commented on src/main/java/meep/parser/CommandParser.java:
```diff
@@ -0,0 +1,34 @@
+     */
+    public static boolean checkEqualCommand(String input, String command) {
+        return input.equalsIgnoreCase(command);
+    }
```
Nice use of comments in the code!

QuekBingHeng commented on src/main/java/meep/parser/Parser.java:
```diff
@@ -0,0 +1,179 @@
+package meep.parser;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import meep.task.Deadline;
+import meep.task.Event;
+import meep.task.TaskList;
+import meep.task.Todo;
+import meep.ui.Ui;
```
Good use of the import statements!

QuekBingHeng commented on src/main/java/meep/parser/Parser.java:
```diff
@@ -0,0 +1,179 @@
+        if (day >= 11 && day <= 13) {
+            return "th";
+        }
+        return switch (day % 10) {
+        case 1 -> "st";
+        case 2 -> "nd";
+        case 3 -> "rd";
+        default -> "th";
+        };
```
Perhaps there is a better way to use switch in the coding standard?

tata32000 replied:
Thanks for the feedback. The IDE suggested using this method instead of the traditional notation, so I'm not sure if this notation complies with the coding standards.

QuekBingHeng commented on src/main/java/meep/task/Storage.java:
```diff
@@ -0,0 +1,128 @@
+            boolean isDone = taskParts[1].equals("1");
+            String taskDescription = taskParts[2];
+            Task task;
+            switch (taskType) {
+            case "T":
+                task = new Todo(taskDescription);
+                if (isDone) {
+                    task.markAsDone();
+                }
+                taskList.addItem(task);
+                break;
+            case "D":
+                task = new Deadline(taskDescription, taskParts[3]);
+                if (isDone) {
+                    task.markAsDone();
+                }
+                taskList.addItem(task);
+                break;
+            case "E":
+                task = new Event(taskDescription, taskParts[3], taskParts[4]);
+                if (isDone) {
+                    task.markAsDone();
+                }
+                taskList.addItem(task);
+                break;
+            default:
+                break;
+            }
```
Great use of switch!

tata32000 replied:
Thank you for your valuable feedback on the code!

QuekBingHeng reviewed (commented):
Overall, great coding structure and coding standard is mostly adhere with small mistakes. Good job!

ArtillerySun commented on src/main/java/meep/parser/CommandParser.java:
```diff
@@ -0,0 +1,34 @@
+ */
+public class CommandParser {
+
+    /**
+     * Checks if the given input string starts with the specified command.
+     * This method is typically used for commands that require additional arguments.
+     *
+     * @param input   The full input string provided by the user.
+     * @param command The command to check against.
+     * @return {@code true} if the input starts with the specified command (case-insensitive), {@code false} otherwise.
+     */
```
It is good to follow the java doc standard！

ArtillerySun commented on src/main/java/meep/parser/Parser.java:
```diff
@@ -0,0 +1,179 @@
+ * The {@code Command} enum represents the various commands that can be issued
+ * by the user to interact with the task management system.
+ */
+enum Command {
+    BYE,
+    MARK,
+    UNMARK,
+    DELETE,
+    DEADLINE,
+    EVENT,
+    TODO,
+    LIST,
+    FIND
+}
```
Good to use enum instead of creating many subclasses!

ArtillerySun commented on src/main/java/meep/parser/Parser.java:
```diff
@@ -0,0 +1,179 @@
+
+        } else if (CommandParser.checkCommand(input, Command.EVENT.toString())) {
+            try {
+                // GitHub Copilot suggested the following code snippet
+                String description = input.split(" ", 2)[1].split(" /from ")[0];
+                // from is between /from and /to
+                String from = formatDate(input.split(" /from ")[1].split(" /to ")[0]);
+                String to = formatDate(input.split(" /to ")[1]);
+                taskList.addItem(new Event(description, from, to));
+                ui.addTask(taskList.getLastTask(), taskList.getSize());
+            } catch (ArrayIndexOutOfBoundsException e) {
+                ui.invalidEventCommand();
+            } catch (DateTimeParseException e) {
+                ui.invalidDateFormat();
+            }
+        } else if (CommandParser.checkCommand(input, Command.TODO.toString())) {
+            try {
+                String description = input.split(" ", 2)[1];
+                taskList.addItem(new Todo(description));
+                ui.addTask(taskList.getLastTask(), taskList.getSize());
+            } catch (ArrayIndexOutOfBoundsException e) {
+                ui.invalidTodoCommand();
+            }
+
+        } else if (CommandParser.checkEqualCommand(input, Command.LIST.toString())) {
+            ui.listTasks(taskList.getList());
+        } else if (CommandParser.checkCommand(input, Command.FIND.toString())) {
+            try {
+                String keyword = input.split(" ", 2)[1];
+                ui.findTasks(taskList.findTasks(keyword));
+            } catch (ArrayIndexOutOfBoundsException e) {
+                ui.invalidFindCommand();
+            }
+        } else {
+            // Any other command will be considered invalid
+            ui.invalidCommand();
+        }
+        return false;
+    }
+}
```
Overall a solid implementation of parser!

ArtillerySun commented on src/main/java/meep/task/Task.java:
```diff
@@ -0,0 +1,66 @@
+ */
+public class Task {
+    private final String description;
+    private boolean isDone;
```
Good boolean variable naming!

ArtillerySun reviewed (commented):
Looks good!
a solid attempt!

tata32000 committed:
Add JavaFX to build.gradle file

tata32000 committed:
Add images resources for GUI

tata32000 committed:
Modify Ui class to return String instead of print

tata32000 committed:
Modify checkCommand method to return response string

tata32000 committed:
Modify EXPECTED.TXT for the new format

tata32000 committed:
Refactor Meep class to support GUI

tata32000 committed:
Add DialogBox class for GUI

tata32000 committed:
Add Launcher class for GUI

tata32000 committed:
Add Main class for GUI

tata32000 committed:
Add fxml files

tata32000 committed:
Add css files

tata32000 committed:
Refactor Main class to use FXML

tata32000 committed:
Add MainWindow class for JavaFX

tata32000 committed:
Modify DialogBox class to use FXML

tata32000 committed:
Remove unused component

tata32000 committed:
Add new photo for the dialog

tata32000 committed:
Add app icon

tata32000 committed:
Remove Duke dialog photo

tata32000 committed:
Change user image

tata32000 committed:
Modify CSS files to change color

tata32000 committed:
Merge pull request #1 from tata32000/branch-Level-10

Branch level 10

tata32000 committed:
Update AI.md

tata32000 committed:
Add assertion in TaskList

tata32000 committed:
Add assertion for day range validation in Parser class

Added an assertion to ensure that the day value is within the valid
range of 1 to 31 before processing. This helps catch errors where
invalid day values might be passed to the parser.

This enhancement improves input validation and ensures that day values
are within an acceptable range, preventing potential issues during
parsing.

tata32000 committed:
Refactor checkCommand and checkEqualCommand methods

Change name of the methods:
- checkCommand to checkCommandWithArgument
- checkEqualCommand to checkCommandWithoutArgument

This is to improve readability and clarity of the code

tata32000 committed:
Refactor createFileIfNotExists method in Storage

tata32000 committed:
Refactor TaskList methods to use Java Streams

Refactored multiple methods in the TaskList class to leverage Java Streams for cleaner, more efficient code:
- getList: Replaced manual iteration with a stream to generate task list with index numbers.
- getSaveFormatList: Streamlined the process of creating the save format by using streams.
- findTasks: Utilized streams to filter tasks based on a keyword search, improving readability and reducing boilerplate.

Additionally, ensured all these methods append a trailing newline after the last task for consistent output formatting.

This refactor improves code clarity and maintains the original functionality while reducing redundancy in iteration.

tata32000 committed:
Update EXPECTED.TXT due to change in find method in TaskList class

tata32000 committed:
Update AI.md

tata32000 committed:
Merge pull request #2 from tata32000/branch-A-Assertions

Branch-A-Assertions

tata32000 committed:
Merge branch 'master' into branch-A-CodeQuality

tata32000 committed:
Merge branch 'master' into branch-A-Streams

tata32000 committed:
Merge pull request #3 from tata32000/branch-A-CodeQuality

Branch-A-CodeQuality

tata32000 committed:
Merge branch 'master' into branch-A-Streams

tata32000 committed:
Merge pull request #4 from tata32000/branch-A-Streams

Branch-A-Streams

tata32000 committed:
Add gradle.yml file for GitHub Actions

tata32000 committed:
Remove IO redirection test for Windows

tata32000 committed:
Merge pull request #5 from tata32000/branch-A-CI

Add gradle.yml file for GitHub Actions

tata32000 committed:
Add help message to Ui class

tata32000 committed:
Add help command to Parser class

tata32000 committed:
Modify text-ui-test to include help command

tata32000 committed:
Merge pull request #6 from tata32000/branch-C-Help

Branch c help

tata32000 committed:
Merge branch 'master' into branch-C-Archive

tata32000 committed:
Add archive messages to Ui Class

tata32000 committed:
Add archiveItem method to TaskList Class

tata32000 committed:
Add archiveTask method to Storage Class

tata32000 committed:
Add archive command to Parser Class

tata32000 committed:
Modify Parser to take in storage for archive

tata32000 committed:
Update .gitignore to ignore archive file

tata32000 committed:
Update test cases to include archive command

tata32000 committed:
Merge pull request #7 from tata32000/branch-C-Archive

Branch c archive

tata32000 committed:
Update the color scheme

tata32000 committed:
Add method to change dialog style in DialogBox class

tata32000 committed:
Modify dialog-box.css to include different labels

tata32000 committed:
Add commandType to Meep.java

tata32000 committed:
Add commandType to Parser.java

tata32000 committed:
Add commandType to MainWindow.java

tata32000 committed:
Remove redundant code in dialog-box.css

tata32000 committed:
Merge pull request #9 from tata32000/branch-A-BetterGui

Branch a better gui

tata32000 committed:
Format main.css

tata32000 committed:
Format Event.java

tata32000 committed:
Add non-existent dates check in Parser.java

tata32000 committed:
Add special characters check in CommandParser.java

tata32000 committed:
Add invalidCharacter method in Ui.java

tata32000 committed:
Merge pull request #10 from tata32000/branch-A-MoreErrorHandling

Branch a more error handling

tata32000 committed:
Add more JUnit tests for Parser class

tata32000 committed:
Merge pull request #11 from tata32000/branch-A-MoreTesting

Add more JUnit tests for Parser class

tata32000 committed:
Refactor checkCommand method in Parser.java

tata32000 committed:
Add javadoc to getCommandType method in Meep.java

tata32000 committed:
Merge pull request #12 from tata32000/branch-A-AiAssisted

Branch a ai assisted

tata32000 committed:
Add Ui.png to /docs

tata32000 committed:
Update README.md for user guide

tata32000 committed:
Merge pull request #13 from tata32000/branch-A-UserGuide

Branch a user guide

tata32000 committed:
Update README.md to include command to run jar file
